pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,143 @@
+// Package timing is used to aggregate timing calls within hotpaths to avoid using
+// repeated statsd calls. The package has a default set that reports at 10 second
+// intervals and can be used directly. If a different behaviour or reporting pattern
+// is desired, a custom Set may be created.
+package timing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet = NewSet()
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Flush reports the default set. It can be useful to call when the program
+// exits prematurely.
+func Flush() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. Use NewSet
+// to create one and Report (or Autoreport) to submit metrics. It is safe for
+// concurrent use.
+type Set struct {
+	mu sync.RWMutex        // guards c
+	c  map[string]*counter // maps names to their aggregates
+}
+
+// NewSet returns a new Set, optionally initialized with the given metric names.
+func NewSet(names ...string) *Set {
+	set := Set{
+		c: make(map[string]*counter, len(names)),
+	}
+	for _, name := range names {
+		set.c[name] = newCounter(name)
+	}
+	return &set
+}
+
+// Autoreport enables autoreporting of the Set at the given interval. It returns a
+// cancellation function.
+func (s *Set) Autoreport(interval time.Duration) (cancelFunc func()) {
+	stop := make(chan struct{})
+	go func() {
+		defer close(stop)
+		tick := time.NewTicker(interval)
+		defer tick.Stop()
+		for {
+			select {
+			case <-tick.C:
+				s.Report()
+			case <-stop:
+				s.Report()
+				return
+			}
+		}
+	}()
+	var once sync.Once // avoid panics
+	return func() {
+		once.Do(func() {
+			stop <- struct{}{}
+			<-stop
+		})
+	}
+}
+
+// Since records the duration for the given metric name as time passed since start.
+// If name does not exist, as defined by NewSet, it creates it.
+func (s *Set) Since(name string, start time.Time) {
+	s.mu.RLock()
+	c, ok := s.c[name]
+	s.mu.RUnlock()
+	if !ok {
+		c = newCounter(name)
+		s.mu.Lock()
+		s.c[name] = c
+		s.mu.Unlock()
+	}
+	ms := time.Since(start) / time.Millisecond
+	c.add(float64(ms))
+}
+
+// Report reports all of the Set's metrics to the statsd client.
+func (s *Set) Report() {
+	s.mu.RLock()
```

Comment:
Assuming `c.flush()` is protected with locking then we don't need locking here as well? Also this should never get called more than the flush interval, if it does that's a problem. Probably want a safety check in `counter.flush` where if the count is zero it doesn't send any metrics. Else the gauges could clobber each other if you do two flushes at the same time, one will be all zeros.

Reply:
> Assuming c.flush() is protected with locking then we don't need locking here as well?

This lock protects reading the map in the `for` loop from potential writes in `Add`

Reply:
Oh right I confused the locks. 👍